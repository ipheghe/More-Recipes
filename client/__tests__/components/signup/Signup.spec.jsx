import expect from 'expect';
import React from 'react';
import { mount, shallow } from 'enzyme';
import sinon from 'sinon';
import render from 'react-test-renderer';
import { HashRouter as Router } from 'react-router-dom';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import ConnectedSignupPage, { PureSignup }
  from '../../../src/components/Signup';


const middleware = [thunk];
const mockStore = configureMockStore(middleware);
const initialState = {
  auth: {
    error: 'Invalid username',
    message: '',
    userData: {},
    authenticated: false,
    categories: []
  }
};

const store = mockStore(initialState);

const state = {
  username: 'okon',
  password: 'abcde',
  fullName: 'Okon Imo',
  mobileNumber: '234702388888',
  email: 'okon@yahoo.com',
  hasErrored: false,
  errorMessage: ''
};

const props = {
  errorMessage: '',
  registerUser: jest.fn(() => Promise.resolve()),
  isAuthenticated: false

};

const event = {
  preventDefault: jest.fn(),
  target: {
    username: 'okon',
    password: 'abcde',
    fullName: 'Okon Imo',
    mobile: '234702388888',
    email: 'okon@yahoo.com'
  }

};

jest.useFakeTimers();

/**
 *@description  setup function to mount component
 *
 * @param { boolean } isAuthenticated
 *
 * @return { * } null
 */
const setup = (isAuthenticated) => {
  props.isAuthenticated = isAuthenticated;
  const mountedWrapper = mount(<Router>
    <ConnectedSignupPage {...props} store={store} />
  </Router>);
  const shallowWrapper = shallow(<PureSignup {...props} />);
  return {
    mountedWrapper,
    shallowWrapper
  };
};

describe('<Signup', () => {
  it('renders without crashing', () => {
    const { mountedWrapper } = setup(false);
    expect(mountedWrapper).toBeDefined();
    expect(mountedWrapper.find('SignupForm').length).toBe(1);
    expect(true).toBe(true);
  });

  it('should match component snapshot', () => {
    const tree = render.create(<Router ><PureSignup {...props} /></Router>);
    expect(tree).toMatchSnapshot();
  });

  it('calls handleChange event', () => {
    sinon.spy(PureSignup.prototype, 'handleChange');
    const { shallowWrapper } = setup(false);
    shallowWrapper.instance().handleChange(event);
    expect(PureSignup.prototype.handleChange.calledOnce).toEqual(true);
  });

  it('calls handleSignup event after Signup button is clicked', () => {
    sinon.spy(PureSignup.prototype, 'handleSignup');
    const { shallowWrapper } = setup(false);
    shallowWrapper.instance().handleSignup(event);
    expect(PureSignup.prototype.handleSignup.calledOnce).toEqual(true);
  });

  it('displays error message if user inputs a null username value', () => {
    sinon.spy(PureSignup.prototype, 'validateFormField');
    const { shallowWrapper } = setup(false);
    state.username = '';
    shallowWrapper.setState(state);
    shallowWrapper.instance().validateFormField();
    expect(shallowWrapper.state().hasErrored).toEqual(true);
    expect(shallowWrapper.state().errorMessage)
      .toEqual('Username must start with a letter and have no spaces.');
    jest.runAllTimers();
    expect(shallowWrapper.state().errorMessage).toEqual('');
  });

  it('displays error message if user inputs a null password value', () => {
    const { shallowWrapper } = setup(false);
    state.username = 'okon';
    state.password = '';
    shallowWrapper.setState(state);
    shallowWrapper.instance().validateFormField();
    expect(shallowWrapper.state().hasErrored).toEqual(true);
    expect(shallowWrapper.state().errorMessage)
      .toEqual('password must contain more than 3 chareacters');
    jest.runAllTimers();
    expect(shallowWrapper.state().errorMessage).toEqual('');
  });

  it(`displays error message if user inputs a 
     fullName value with length less than 4`, () => {
      const { shallowWrapper } = setup(false);
      state.password = 'abcde';
      state.fullName = 'oko';
      shallowWrapper.setState(state);
      shallowWrapper.instance().validateFormField();
      expect(shallowWrapper.state().hasErrored).toEqual(true);
      expect(shallowWrapper.state().errorMessage)
        .toEqual('fullname must contain more than 3 chareacters');
      jest.runAllTimers();
      expect(shallowWrapper.state().errorMessage).toEqual('');
    });

  it('displays error message if user inputs an invalid mobile number', () => {
    const { shallowWrapper } = setup(false);
    state.mobileNumber = '';
    state.fullName = 'okon Akem';
    shallowWrapper.setState(state);
    shallowWrapper.instance().validateFormField();
    expect(shallowWrapper.state().hasErrored).toEqual(true);
    expect(shallowWrapper.state().errorMessage)
      .toEqual('mobile number must contain only numbers');
    jest.runAllTimers();
    expect(shallowWrapper.state().errorMessage).toEqual('');
  });

  it('displays error message if user inputs an invalid email', () => {
    const { shallowWrapper } = setup(false);
    state.mobileNumber = '234702388888';
    state.email = 'okon@ya';
    shallowWrapper.setState(state);
    shallowWrapper.instance().validateFormField();
    expect(shallowWrapper.state().hasErrored).toEqual(true);
    expect(shallowWrapper.state().errorMessage)
      .toEqual('Invalid Email Address');
    jest.runAllTimers();
    expect(shallowWrapper.state().errorMessage).toEqual('');
  });

  it(`dispatches registerUser action after 
     validatiing input fields successfully`, () => {
      const { shallowWrapper } = setup(false);
      state.email = 'okon@yahoo.com';
      shallowWrapper.setState(state);
      shallowWrapper.instance().validateFormField();
      expect(PureSignup.prototype.validateFormField.calledOnce).toEqual(false);
      expect(shallowWrapper.state().password).toEqual('abcde');
    });

  it('redirects to dashboard page if user is authenticated', () => {
    const { shallowWrapper } = setup(true);
    expect(shallowWrapper.instance().props.isAuthenticated).toEqual(true);
  });
});
