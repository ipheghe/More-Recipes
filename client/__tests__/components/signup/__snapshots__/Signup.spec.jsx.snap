// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Signup should match component snapshot 1`] = `
<div>
  <div
    className="login-background"
  >
    <div
      className="container"
    >
      <div
        className="row landing"
      >
        <section
          className="col-md-6 headline"
        >
          <h1>
            Welcome to More recipes
          </h1>
          <h4>
            Share your recipe ideas with the world
          </h4>
          <br />
          <br />
          <h2>
            Sign Up and start sharing recipes
          </h2>
        </section>
        <section
          className="col-md-1"
        />
        <section
          className="col-md-5 account"
        >
          <div>
            <form
              className="reg-form"
            >
              <h3
                className="login-form-boxx"
              >
                Registration Form
              </h3>
              <br />
              <div
                className="form-group"
              >
                <label
                  htmlFor="enterUsername"
                >
                  Username:
                </label>
                <div
                  className="input-group"
                >
                  <span
                    className="input-group-addon"
                  >
                    <i
                      className="fa fa-user-secret"
                    />
                  </span>
                  <input
                    className="form-control"
                    id="username"
                    name="username"
                    onChange={[Function]}
                    placeholder="e.g john112"
                    required={true}
                    type="text"
                    value=""
                  />
                </div>
              </div>
              <div
                className="form-group"
              >
                <label
                  htmlFor="enterFullName"
                >
                  Full Name:
                </label>
                <div
                  className="input-group"
                >
                  <span
                    className="input-group-addon"
                  >
                    <i
                      className="fa fa-user"
                    />
                  </span>
                  <input
                    className="form-control"
                    id="fullName"
                    name="fullName"
                    onChange={[Function]}
                    placeholder="e.g John James"
                    required={true}
                    type="text"
                    value=""
                  />
                </div>
              </div>
              <div
                className="form-group"
              >
                <label
                  htmlFor="enterMobile"
                >
                  Mobile:
                </label>
                <div
                  className="input-group"
                >
                  <span
                    className="input-group-addon"
                  >
                    <i
                      className="fa fa-mobile"
                    />
                  </span>
                  <input
                    className="form-control"
                    id="mobileNumber"
                    name="mobileNumber"
                    onChange={[Function]}
                    placeholder="e.g 08034562838"
                    required={true}
                    type="number"
                    value=""
                  />
                </div>
              </div>
              <div
                className="form-group"
              >
                <label
                  htmlFor="enterEmail"
                >
                  Email address:
                </label>
                <div
                  className="input-group"
                >
                  <span
                    className="input-group-addon"
                  >
                    <i
                      className="fa fa-envelope"
                    />
                  </span>
                  <input
                    className="form-control"
                    id="email"
                    name="email"
                    onChange={[Function]}
                    placeholder="e.g john@yahoo.com"
                    required={true}
                    type="email"
                    value=""
                  />
                </div>
              </div>
              <div
                className="form-group"
              >
                <label
                  htmlFor="enterPassword"
                >
                  Password:
                </label>
                <div
                  className="input-group"
                >
                  <span
                    className="input-group-addon"
                  >
                    <i
                      className="fa fa-unlock-alt"
                    />
                  </span>
                  <input
                    className="form-control"
                    id="password"
                    name="password"
                    onChange={[Function]}
                    placeholder=""
                    required={true}
                    type="password"
                    value=""
                  />
                </div>
              </div>
              <div
                className="row"
              >
                <div
                  className="col-md-6"
                >
                  <a
                    className="have-account"
                    href="#login"
                  >
                    Already have an account?
                  </a>
                </div>
              </div>
              <div
                className="login-buttons"
              >
                <div>
                  <a>
                    <button
                      className="btn btn-block btn-success"
                      id="signup"
                      onClick={[Function]}
                    >
                      Sign Up
                    </button>
                  </a>
                </div>
              </div>
            </form>
          </div>
        </section>
        <br />
      </div>
    </div>
  </div>
</div>
`;
